app/Http/Controllers/Shop/CheckoutController.php

```diff
@@ -97,16 +97,25 @@ public function process(ProcessRequest $request): CheckoutResource
 			$request->additional_data ?? []
 		);
 
-		// If we have a success directly without redirection mark order as completed
 		if ($result->is_success && !$result->is_redirect) {
-			OrderCompleted::dispatchIf(Configs::getValueAsBool('webshop_auto_fulfill_enabled'), $order->id);
+			// This is a success we now need to complete the order.
```

Comment:
This way the OrderCompleted event is only dispatched in the finalize step, not in the processing step.
Furthermore that clarifies that we have complete_url and redirect_url, showing different intent.